### PR TITLE
Add `poetry`, `tox`, `Dockerfile`, and `buildspec.test` configurations

### PR DESCRIPTION
**Pull request by yusufsheiqh**
## Context

- Add `poetry` configuration
- Add `tox` configuration
  - Currently added support for `3.9`, `3.10`, and `3.11`.
- Add `Dockerfile`
  - Use in CodeBuild project as we need access to multiple versions of Python runtimes for `tox`.
- Update `buildspec` configurations
  - Run `tox` in the CodeBuild project.
  - Run tests locally via the pre-commit hook when committing and also via `poetry run test`.
- Add `--version` flag
- Update documentation

## Additional information

The current CodeBuild project will be replaced with the updated CodeBuild project in a subsequent PR. The build badge will also be updated in a subsequent PR.

Note: All developers should make sure they have run `pre-commit` installed and set up by following the documentation.

JIRA ticket - https://uktrade.atlassian.net/browse/DBTP-135

### CODE REVIEW

File: tests/test_copilot_helper.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+from click.testing import CliRunner
+
+from copilot_helper import cli
+
+
+class TestCopilotHelperCli:
+    def test_check_version(self):
+        runner = CliRunner()
+
+        result = runner.invoke(cli, ["--version"])
```
Comment (WillGibson) [inline tests/test_copilot_helper.py:12]: You could just inline the runner...
```suggestion
        result = CliRunner().invoke(cli, ["--version"])
```

**Review by WillGibson — APPROVED** [review]
Nice. I've not seen Poetry before. It looks sweet.

**Review by gabelton — APPROVED** [review]

File: tests/test_copilot_helper.py (hunk 1 above)
Comment (yusufsheiqh) [reply tests/test_copilot_helper.py:12]: True dat. I thought I would be writing moar tests, but decided this was enough for now.

Comment (yusufsheiqh) [reply tests/test_copilot_helper.py]: Resolved in https://github.com/uktrade/copilot-tools/pull/51/commits/ddb762b327dfdc70273dd134078d043c96db5d46.

**Pull request merged**